File: test/validate.test.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { git } from "../src/git.js";
import { initialize } from "../src/init.js";
import { validateRepository } from "../src/validate.js";

const temporaryDirectories: string[] = [];

afterEach(async () => {
  await Promise.all(
    temporaryDirectories
      .splice(0)
      .map((directory) => fs.rm(directory, { recursive: true, force: true })),
  );
});

describe("repository validation", () => {
  it("requires the release lockfile to exist in Git", async () => {
    const root = await fs.mkdtemp(path.join(os.tmpdir(), "smf-validate-"));
    temporaryDirectories.push(root);
    await fs.mkdir(path.join(root, "ios"));
    await fs.writeFile(
      path.join(root, "pubspec.yaml"),
      "name: example\nversion: 1.0.0+1\n",
    );
    await git(root, ["init", "-b", "main"]);
    await git(root, ["config", "user.name", "Test"]);
    await git(root, ["config", "user.email", "test@example.com"]);
    await git(root, ["add", "."]);
    await git(root, ["commit", "-m", "chore: bootstrap"]);
    await initialize({ root, bundleId: "dev.example.app" });
    await git(root, ["add", "."]);
    await git(root, ["commit", "-m", "chore: configure releases"]);

    await expect(validateRepository(root)).rejects.toThrow(
      /No committed pubspec\.lock/,
    );
    await fs.writeFile(path.join(root, "pubspec.lock"), "# fixture\n");
    await expect(validateRepository(root)).rejects.toThrow(
      /pubspec\.lock must be committed/,
    );
    await git(root, ["add", "pubspec.lock"]);
    await git(root, ["commit", "-m", "chore: lock dependencies"]);
    await expect(validateRepository(root)).resolves.toBeUndefined();
  });
});

File: test/validate_test.dart
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:smf/smf.dart';
import 'package:test/test.dart';

void main() {
  test('repository validation requires the release lockfile in Git', () async {
    final root = await Directory.systemTemp.createTemp('smf-validate-');
    addTearDown(() => root.delete(recursive: true));
    await Directory(p.join(root.path, 'ios')).create();
    await File(
      p.join(root.path, 'pubspec.yaml'),
    ).writeAsString('name: example\nversion: 1.0.0+1\n');
    await git(root.path, const <String>['init', '-b', 'main']);
    await git(root.path, const <String>['config', 'user.name', 'Test']);
    await git(root.path, const <String>[
      'config',
      'user.email',
      'test@example.com',
    ]);
    await git(root.path, const <String>['add', '.']);
    await git(root.path, const <String>['commit', '-m', 'chore: bootstrap']);
    await initialize(
      InitOptions(appRoot: root.path, bundleId: 'dev.example.app'),
    );
    await git(root.path, const <String>['add', '.']);
    await git(root.path, const <String>[
      'commit',
      '-m',
      'chore: configure releases',
    ]);

    await expectLater(
      validateRepository(root.path),
      throwsA(
        isA<SmfError>().having(
          (SmfError error) => error.message,
          'message',
          contains('No committed pubspec.lock'),
        ),
      ),
    );
    await File(p.join(root.path, 'pubspec.lock')).writeAsString('# fixture\n');
    await expectLater(
      validateRepository(root.path),
      throwsA(
        isA<SmfError>().having(
          (SmfError error) => error.message,
          'message',
          contains('pubspec.lock must be committed'),
        ),
      ),
    );
    await git(root.path, const <String>['add', 'pubspec.lock']);
    await git(root.path, const <String>[
      'commit',
      '-m',
      'chore: lock dependencies',
    ]);
    await validateRepository(root.path);
  });
}
